src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py

```diff
@@ -1246,41 +1271,52 @@ def _visit_reduce(self, node: itir.FunCall):
                 """
-                The neighbors builtin returns an array of booleans in case the connectivity table
-                contains skip values. These boolean values indicate whether the neighbor value is present or not,
-                and are used below to construct an if/else branch to bypass the lambda call for neighbor skip values.
+                The neighbors builtin returns an array of booleans in case the connectivity table contains skip values.
+                These booleans indicate whether the neighbor is present or not, and are used in a tasklet to select
+                the result of field access or the identity value, respectively.
                 If the neighbor table has full connectivity (no skip values by type definition), the input_valid node
-                is not built, and the construction of the if/else branch below is also skipped.
+                is not built, and the construction of the select tasklet below is also skipped.
                 """
-                input_args.append(input_valid[0])
-                input_valid_node = input_valid[0].value
+                input_args.append(input_valid_args[0])
+                input_valid_node = input_valid_args[0].value
+                lambda_output_node = inner_outputs[0].value
                 # add input connector to nested sdfg
-                input_mapping["is_valid"] = create_memlet_at(input_valid_node.data, nreduce_index)
-                # check neighbor validity on if/else inter-state edge
-                start_state = lambda_context.body.add_state("start", is_start_block=True)
-                skip_neighbor_state = lambda_context.body.add_state("skip_neighbor")
-                skip_neighbor_state.add_edge(
-                    skip_neighbor_state.add_tasklet(
-                        "identity", {}, {"val"}, f"val = {reduce_identity}"
-                    ),
-                    "val",
-                    skip_neighbor_state.add_access(inner_outputs[0].value.data),
+                lambda_context.body.add_scalar("_valid_neighbor", dace.dtypes.bool)
+                input_mapping["_valid_neighbor"] = create_memlet_at(
+                    input_valid_node.data, nreduce_index
+                )
+                # add select tasklet before writing to output node
+                output_edge = lambda_context.state.in_edges(lambda_output_node)[0]
+                select_tasklet = lambda_context.state.add_tasklet(
+                    "neighbor_select",
+                    {"_inp", "_valid"},
+                    {"_out"},
+                    f"_out = _inp if _valid else {reduce_identity}",
+                )
+                lambda_context.state.add_edge(
+                    output_edge.src,
                     None,
-                    dace.Memlet(data=inner_outputs[0].value.data, subset="0"),
+                    select_tasklet,
+                    "_inp",
+                    dace.Memlet(data=output_edge.src.data, subset="0"),
```

Comment:
I do not understand why you have here `subset="0"`, doesn't this mean that you always read the first element of the array or I am wrong?